File: tests/sample.rb
# frozen_string_literal: true

# User management service for handling user operations
# Provides CRUD operations and validation

require 'date'

# User model with validation
class User
  attr_accessor :id, :name, :email, :role, :active, :metadata

  VALID_ROLES = %i[admin user guest].freeze
  EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i.freeze

  def initialize(id:, name:, email:, role: :user, active: true, metadata: {})
    @id = id
    @name = name
    @email = validate_email!(email)
    @role = validate_role!(role)
    @active = active
    @metadata = metadata
  end

  def admin?
    @role == :admin
  end

  def to_s
    "User ##{id}: #{name} (#{email}) - #{role}"
  end

  private

  def validate_email!(email)
    raise ArgumentError, "Invalid email: #{email}" unless email.match?(EMAIL_REGEX)
    email
  end

  def validate_role!(role)
    role = role.to_sym
    raise ArgumentError, "Invalid role: #{role}" unless VALID_ROLES.include?(role)
    role
  end
end

# Service for managing users
class UserService
  MAX_USERS = 1000

  def initialize
    @users = {}
    initialize_users
  end

  # Get user by ID
  # @param id [Integer] User's ID
  # @return [User, nil] User or nil if not found
  def get_user(id)
    @users[id]
  end

  # Find users matching criteria
  # @param filters [Hash] Filter criteria
  # @return [Array<User>] Matching users
  def find_users(**filters)
    @users.values.select do |user|
      filters.all? { |key, value| user.send(key) == value }
    end
  end

  # Create a new user
  # @return [User] The created user
  # @raise [StandardError] If max users reached
  def create_user(name:, email:, **options)
    raise StandardError, 'Maximum user limit reached' if @users.size >= MAX_USERS

    id = (@users.keys.max || 0) + 1
    metadata = { created_at: DateTime.now.iso8601 }
    
    user = User.new(
      id: id,
      name: name,
      email: email,
      metadata: metadata,
      **options
    )
    
    @users[id] = user
    user
  rescue ArgumentError => e
    puts "Error creating user: #{e.message}"
    nil
  end

  # Get all active users
  def active_users
    find_users(active: true)
  end

  private

  def initialize_users
    default_users = [
      { id: 1, name: 'Alice', email: 'alice@example.com', role: :admin },
      { id: 2, name: 'Bob', email: 'bob@example.com', role: :user },
      { id: 3, name: 'Charlie', email: 'charlie@example.com', active: false }
    ]

    default_users.each do |attrs|
      user = User.new(**attrs)
      @users[user.id] = user
    end
  end
end

# Example usage
if __FILE__ == $PROGRAM_NAME
  service = UserService.new

  # Find active users
  active = service.active_users
  puts "Found #{active.size} active users"

  # Get specific user
  user = service.get_user(1)
  puts user if user

  # Create new user
  new_user = service.create_user(
    name: 'Diana',
    email: 'diana@example.com',
    role: :user
  )
  puts "Created: #{new_user}" if new_user

  # Find all admins
  admins = service.find_users(role: :admin)
  puts "\nAdmins:"
  admins.each { |admin| puts "  - #{admin.name}" }
end

File: tests/sample.py
"""User management module for handling authentication and user data"""

from dataclasses import dataclass, field
from typing import Optional, List, Dict, Any
from datetime import datetime
import re


@dataclass
class User:
    """Represents a user in the system"""
    id: int
    name: str
    email: str
    role: str = 'user'
    active: bool = True
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self._validate_email(self.email):
            raise ValueError(f"Invalid email: {self.email}")

    @staticmethod
    def _validate_email(email: str) -> bool:
        """Validates email format using regex"""
        pattern = r'^[\w\.-]+@[\w\.-]+\.\w+$'
        return re.match(pattern, email) is not None


class UserService:
    """Service for managing user operations"""
    
    MAX_USERS = 1000

    def __init__(self):
        self.users: Dict[int, User] = {}
        self._initialize_users()

    def _initialize_users(self) -> None:
        """Initialize with default users"""
        default_users = [
            User(1, 'Alice', 'alice@example.com', role='admin'),
            User(2, 'Bob', 'bob@example.com'),
            User(3, 'Charlie', 'charlie@example.com', active=False)
        ]
        for user in default_users:
            self.users[user.id] = user

    def get_user(self, user_id: int) -> Optional[User]:
        """Retrieve user by ID"""
        return self.users.get(user_id)

    def find_users(self, **filters) -> List[User]:
        """Find users matching the given filters"""
        results = []
        
        for user in self.users.values():
            match = True
            for key, value in filters.items():
                if not hasattr(user, key) or getattr(user, key) != value:
                    match = False
                    break
            if match:
                results.append(user)
        
        return results

    def create_user(self, name: str, email: str, **kwargs) -> User:
        """Create a new user"""
        user_id = max(self.users.keys(), default=0) + 1
        new_user = User(
            id=user_id,
            name=name,
            email=email,
            metadata={'created_at': datetime.now().isoformat()},
            **kwargs
        )
        self.users[user_id] = new_user
        return new_user


# Example usage
if __name__ == '__main__':
    service = UserService()
    
    # Find all active users
    active_users = service.find_users(active=True)
    print(f"Found {len(active_users)} active users")
    
    # Get specific user
    user = service.get_user(1)
    if user:
        print(f"User: {user.name} ({user.email}) - Role: {user.role}")
    
    # Create new user
    new_user = service.create_user('Diana', 'diana@example.com', role='user')
    print(f"Created user: {new_user.name} with ID {new_user.id}")
